### PR TITLE
refactor: introduce redux toolkit

### PR DESCRIPTION
**Pull request by justinpark**
### SUMMARY
Following up the discussion of #23257,
this commit introduces the redux-toolkit as a prerequisite for rdk(redux toolkit) query.

In order to consolidate the multiple state management solution, superset team decided to leverage the redux with toolkit in order to organize the existing redux state.
Redux toolkit also includes the rdk query package which can replace the existing react-query hooks (which is a similar solution to minimize the state of async api request) with a single solution.

This commit also use the configureStore method of the @reduxjs/toolkit package, which replaces createStore since Redux Toolkit is our (redux official channel) recommended approach for writing Redux logic today, including store setup, reducers, data fetching, and more.

After this commit merged, we can replace existing react-query by rdk-query too. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

N/A

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

cc: @michael-s-molina @ktmud

### CODE REVIEW

File: superset-frontend/spec/fixtures/mockStore.js
Comment (michael-s-molina) [inline superset-frontend/spec/fixtures/mockStore.js:39]:
```suggestion
    disableDebugger: true,
```

File: superset-frontend/spec/fixtures/mockStore.js
Comment (michael-s-molina) [inline superset-frontend/spec/fixtures/mockStore.js:47]:
```suggestion
    disableDebugger: true,
```

File: superset-frontend/spec/fixtures/mockStore.js
Comment (michael-s-molina) [inline superset-frontend/spec/fixtures/mockStore.js:57]:
```suggestion
    disableDebugger: true,
```

File: superset-frontend/spec/fixtures/mockStore.js
Comment (michael-s-molina) [inline superset-frontend/spec/fixtures/mockStore.js:102]:
```suggestion
    disableDebugger: true,
```

File: superset-frontend/spec/fixtures/mockStore.js
Comment (michael-s-molina) [inline superset-frontend/spec/fixtures/mockStore.js:136]:
```suggestion
    disableDebugger: true,
```

File: superset-frontend/src/views/menu.tsx
Comment (michael-s-molina) [inline superset-frontend/src/views/menu.tsx:36]:
```suggestion
const store = setupStore({ disableDebugger: true });
```

File: superset-frontend/src/views/store.ts
Comment (michael-s-molina) [inline superset-frontend/src/views/store.ts:124]:
```suggestion
  disableDebugger = false,
```

File: superset-frontend/src/views/store.ts
Comment (michael-s-molina) [inline superset-frontend/src/views/store.ts:129]:
```suggestion
  disableDebugger?: boolean;
```

File: superset-frontend/src/views/store.ts
Comment (michael-s-molina) [inline superset-frontend/src/views/store.ts:139]:
```suggestion
    devTools: process.env.WEBPACK_MODE === 'development' && !disableDebugger,
```

**Comment by michael-s-molina** [comment]
Thank you so much @justinpark for this PR!

I'm tagging some folks here to help with testing and also for awareness given that Redux Toolkit improves the way to access the store and can really help with new features.

@villebro @geido @kgabryje @eschutho @betodealmeida

File: superset-frontend/spec/fixtures/mockStore.js
Comment (justinpark) [reply superset-frontend/spec/fixtures/mockStore.js:33]: good catch

**Comment by michael-s-molina** [comment]
For those reviewing the PR, keep in mind that [getDefaultMiddleware](https://redux-toolkit.js.org/api/getDefaultMiddleware#included-default-middleware) includes the [immutabilityMiddleware](https://redux-toolkit.js.org/api/immutabilityMiddleware) and [serializabilityMiddleware](https://redux-toolkit.js.org/api/serializabilityMiddleware) by default in development environments (it does not include them in production). 

This will produce warnings and errors such as:

Screenshot 2023-03-24 at 15 08 09

Screenshot 2023-03-24 at 14 58 01

These errors are expected, and a great addition to our quality checks, and we should work to resolve them incrementally.

**Comment by michael-s-molina** [comment]
@justinpark The immutability middleware is finding problems in many places.

https://user-images.githubusercontent.com/70410625/227622214-4da88861-7051-4f22-8e4e-21477955593e.mov

<br>

I see two possible solutions:
- Try to fix the errors in this PR, so that when it's merged, there's no disruption
- Fix the errors incrementally
  - Disable the immutability middleware by default for now
  - Work on fixes for the problems (by enabling the middleware locally when working on PRs)
  - After all the problems are fixed, enable it by default again

**Comment by justinpark** [comment]
> justinpark The immutability middleware is finding problems in many places.
> 
> 
> I see two possible solutions:
> 
> * Try to fix the errors in this PR, so that when it's merged, there's no disruption
> * Fix the errors incrementally
>   
>   * Disable the immutability middleware by default for now
>   * Work on fixes for the problems (by enabling the middleware locally when working on PRs)
>   * After all the problems are fixed, enable it by default again

@michael-s-molina thanks for the detail testing and catching the side effects.

In order to properly use redux-toolkit going forward, I believe it is more appropriate to first address the current errors rather than excluding additional middleware (immutability, serializability). This is because future planned features, such as rdk query, may be dependent on such middlewares.

This has raised questions about whether it is the right direction to use the **redux-toolkit query** before proceeding with bug fixes. The redux toolkit, which is being introduced to simplify the redux structure, seems to be causing more complex dependency issues like this.
Additionally, the redux toolkit query, which was introduced to replace react-query, not only has these dependency issues, but also increases the complexity of writing test code, as discovered during the current react-query migration work. (I will follow up more detail about the complexity in test codes next week)

**Comment by michael-s-molina** [comment]
> In order to properly use redux-toolkit going forward, I believe it is more appropriate to first address the current errors rather than excluding additional middleware (immutability, serializability).

No problem. I can help you test and identify the errors, and submit fixes. Let's keep this PR open and address the issues first then.

> The redux toolkit, which is being introduced to simplify the redux structure, seems to be causing more complex dependency issues like this.

@justinpark RTK is actually pointing out bugs that currently exist in our Redux use but are not noticed. I really like the fact that now we can see these errors and actually fix them. This will improve quality.

> Additionally, the redux toolkit query, which was introduced to replace react-query, not only has these dependency issues, but also increases the complexity of writing test code, as discovered during the current react-query migration work.

Even if the complexity is a little higher, I would vote for library consistency here. Over the years we had a proliferation of libraries that many times offered the same functionality or were not consistent with each other. That resulted in many migrations over the years and wasted resources. I think it's an acceptable trade off.

**Comment by michael-s-molina** [comment]
@justinpark I started submitting PRs to fix state mutations and help us merge this PR. I'm referencing this PR when submitting the fixes so you can follow them.

**Comment by justinpark** [comment]
> > In order to properly use redux-toolkit going forward, I believe it is more appropriate to first address the current errors rather than excluding additional middleware (immutability, serializability).
> 
> No problem. I can help you test and identify the errors, and submit fixes. Let's keep this PR open and address the issues first then.
> 
> > The redux toolkit, which is being introduced to simplify the redux structure, seems to be causing more complex dependency issues like this.
> 
> justinpark RTK is actually pointing out bugs that currently exist in our Redux use but are not noticed. I really like the fact that now we can see these errors and actually fix them. This will improve quality.
> 
> > Additionally, the redux toolkit query, which was introduced to replace react-query, not only has these dependency issues, but also increases the complexity of writing test code, as discovered during the current react-query migration work.
> 
> Even if the complexity is a little higher, I would vote for library consistency here. Over the years we had a proliferation of libraries that many times offered the same functionality or were not consistent with each other. That resulted in many migrations over the years and wasted resources. I think it's an acceptable trade off.

Got your point. Okay I'll go ahead of the rtk query migration.

**Comment by justinpark** [comment]
@michael-s-molina I rebase the branch including #23522
could you verify the issue resolved?

File: superset-frontend/src/views/store.ts
Comment (michael-s-molina) [inline superset-frontend/src/views/store.ts:145]:
```suggestion
      getDefaultMiddleware({
        getDefaultMiddleware({
        immutableCheck: {
          warnAfter: 200,
        },
        serializableCheck: {
          // Ignores AbortController instances
          ignoredActionPaths: [/queryController/g],
          ignoredPaths: [/queryController/g],
          warnAfter: 200,
        },
      }).concat(logger),
```

**Comment by justinpark** [comment]
@michael-s-molina Since we changed the violation notice of immutability and serializability from error to warn, your iterations won't be a hard blocker.

**Comment by michael-s-molina** [comment]
> @michael-s-molina Since we changed the violation notice of immutability and serializability from error to warn, your iterations won't be a hard blocker.

@justinpark `warnAfter` does not change from error to warning. It just adjusts the warnings threshold. State changes will still throw errors and break functionality. I'm working on Iteration 3 which will probably be enough to merge this PR given that all major flows will be preserved.

**Comment by justinpark** [comment]
> justinpark `warnAfter` does not change from error to warning. It just adjusts the warnings threshold. State changes will still throw errors and break functionality. I'm working on Iteration 3 which will probably be enough to merge this PR given that all major flows will be preserved.

Got it. Thanks for confirm. Will wait until iteration is merged

**Comment by justinpark** [comment]
rebased master including iteration 2 (#23535)

**Review by rusackas — APPROVED** [review]
LGTM! Seems cleaner and safer already :D

**Comment by rusackas** [comment]
Slapping a `hold!` label on this since it sounds like https://github.com/apache/superset/pull/23637 should be merged first.

**Review by michael-s-molina — APPROVED** [review]
LGTM

**Comment by justinpark** [comment]
@rusackas since #23637 merged, can you release the hold?

**Pull request merged**
